subsystem/src/subsystem.rs

```diff
@@ -108,58 +116,82 @@ impl<T> Clone for Handle<T> {
 
 #[derive(Debug, Ord, PartialOrd, PartialEq, Eq, Clone, Copy, thiserror::Error)]
 pub enum CallError {
-    #[error("Callee subsystem has terminated")]
-    SubsystemDead,
+    #[error("Call submission failed")]
+    SubmissionFailed,
+    #[error("Result retrieval failed")]
+    ResultFetchFailed,
+}
+
+pub struct CallResult<T>(Result<CallResponse<T>, CallError>);
+
+impl<T> CallResult<T> {
+    /// Get the corresponding [`CallResponse`], with the opportunity to handle errors at send time.
+    pub fn response(self) -> Result<CallResponse<T>, CallError> {
+        self.0
+    }
+}
+
+impl<T> std::future::Future for CallResult<T> {
+    type Output = Result<T, CallError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> task::Poll<Self::Output> {
+        self.0.as_mut().map_or_else(
+            |err| task::Poll::Ready(Err(*err)),
+            |res| Pin::new(res).poll(cx),
+        )
+    }
 }
 
 impl<T: Send + 'static> Handle<T> {
     /// Crate a new subsystem handle.
-    pub(crate) fn new(action_tx: mpsc::Sender<Action<T, ()>>) -> Self {
+    pub(crate) fn new(action_tx: ActionSender<T>) -> Self {
         Self { action_tx }
     }
 
-    /// Dispatch an async function call to the subsystem
-    pub async fn call_async_mut<R: Send + 'static>(
+    /// Call an async procedure to the subsystem. Result has to be await-ed explicitly
+    pub fn call_async_mut<R: Send + 'static>(
         &self,
         func: impl for<'a> FnOnce(&'a mut T) -> BoxFuture<'a, R> + Send + 'static,
-    ) -> Result<R, CallError> {
+    ) -> CallResult<R> {
         let (rtx, rrx) = oneshot::channel::<R>();
 
-        self.action_tx
+        let res = self
+            .action_tx
             .send(Box::new(move |subsys| {
```

Comment:
Please see if you can get rid of this Box, since everything is a generic. Doesn't have to be in this PR.

Reply:
I don't think it's easily achievable. The channel endpoint has type `Sender<Box<dyn Fn(_)>>`. It has to be a box since the `Fn` closure may a different type for each invocation of the `call` procedure.